src/main/java/at/hannibal2/skyhanni/features/misc/pathfind/NavigationHelper.kt

```diff
@@ -0,0 +1,109 @@
+        GraphNodeTag.MINES_EMISSARY,
+        GraphNodeTag.CRIMSON_MINIBOSS,
+    )
+
+    fun onCommand(args: Array<String>) {
+        SkyHanniMod.coroutineScope.launch {
+            asyncTest(args)
+        }
+    }
+
+    private fun asyncTest(args: Array<String>) {
+        val searchTerm = args.joinToString(" ").lowercase()
+        val distances = calculateDistances(searchTerm)
+        val names = calculateNames(distances)
+
+        val text = mutableListOf<IChatComponent>()
+        text.add(Text.createDivider())
+        text.add("§7Found ${names.size} locations ($searchTerm)".asComponent().center())
+        val goBack = {
+            onCommand(searchTerm.split(" ").toTypedArray())
+            IslandGraphs.stop()
+        }
+        // TODO dont show a too long list, add pages
+        for ((name, node) in names) {
+            val distance = distances[node]!!.round(1)
+            val component = "$name §e$distance".asComponent()
+            component.onClick {
+                IslandGraphs.pathFind(node.position)
+                sendNavigateMessage(name, goBack)
+            }
+            val tag = node.tags.first { it in allowedTags }
+            // TODO include most closest area, if this is no area (type in area = forger in forge)
+            component.hover = ("§eClick to start navigating to\n" + "§7Type: §r${tag.displayName}\n" + "§7Distance: §e$distance blocks").asComponent()
+            text.add(component)
+        }
+        text.add(Text.createDivider())
+        Text.multiline(text).send(NAVIGATION_CHAT_ID)
+    }
+
+    private fun sendNavigateMessage(name: String, goBack: () -> Unit) {
+        val componentText = "§7Navigating to §r$name".asComponent()
+        componentText.onClick(onClick = goBack)
+        componentText.send(NAVIGATION_CHAT_ID)
+    }
+
+    private fun calculateNames(distances: Map<GraphNode, Double>): MutableMap<String, GraphNode> {
+        val names = mutableMapOf<String, GraphNode>()
+        for (node in distances.sorted().keys) {
+            val tag = node.tags.first { it in allowedTags }
+            val name = "${node.name} §7(${tag.displayName}§7)"
+            if (name in names) continue
+            names[name] = node
+        }
+        return names
+    }
+
+    private fun calculateDistances(
+        searchTerm: String,
+    ): Map<GraphNode, Double> {
+        val grapth = IslandGraphs.currentIslandGraph ?: return emptyMap()
+        val closedNote = IslandGraphs.closedNote ?: return emptyMap()
+
+        val nodes = grapth.nodes
+        val distances = mutableMapOf<GraphNode, Double>()
+        for (node in nodes) {
+            val name = node.name ?: continue
+            if (!node.tags.any { it in allowedTags }) continue
+            if (name.lowercase().contains(searchTerm)) {
+                distances[node] = grapth.findShortestDistance(closedNote, node)
+            }
+            val remainingTags = node.tags.filter { it in allowedTags }
```

Comment:
while we are checking for different stuff here we still end up filtering over the tags twice, this value could be stored and then we use .isEmpty or something instead (and then can look at the size as well to check if its 1)